### ISSUE TITLE
Command to reproduce OGB arxiv results?

### ISSUE THREAD

**Issue by Chillee** [dmlc/dgl#2314]
Hi

I just wanted to verify that the command to reproduce the OGB results from (https://github.com/Espylapiza/dgl/tree/master/examples/pytorch/ogb/ogbn-arxiv) is correct.

The command is `python3 gat.py --use-norm --use-labels --n-label-iters=1 --no-attn-dst --edge-drop=0.3 --input-drop=0.25`?

In my 10 runs, I got `73.92`, `73.83`,`74.00`,`73.97`,  `73.88`, `73.88`, `73.81`, `73.84`, `74.03`, `73.72`, `73.81`, for an average of `73.88 +- 0.09`. Considering that only 2 of these runs matched the reported accuracy, and that the reported accuracy is not within a standard deviation of what I got, I suspect I'm not running the right command or perhaps have the wrong version. Maybe a different number of epochs?

I'm using PyTorch `1.6.0` and `dgl-cuda10.2` version `0.5.2`.

I'll try rerunning to see if it's just variance, but thought I'd ask in the meantime.

**Comment by yzh119** [comment]
Hi Horace, I run for another ten runs, and the result is as follows:
```
Runned 10 times
Val Accs: [0.7511661465149837, 0.7504949830531226, 0.7500922849760059, 0.7499916104567267, 0.7502265176683781, 0.7509647974764254, 0.7514681700728212, 0.7511661465149837, 0.7513674955535421, 0.7511997046880767]
Test Accs: [0.7411476657819476, 0.7387404069707632, 0.7398514494990022, 0.7359628006501656, 0.7402012221467811, 0.737567639857622, 0.7389667304487377, 0.7412299652284838, 0.7392136287883464, 0.7386581075242269]
Average val accuracy: 0.7508137856975066 ± 0.0005289538839207193
Average test accuracy: 0.7391539616896077 ± 0.001517344084378135
Number of params: 1441580
```

As the initial seed is not fixed I think we can not guarantee reproducibility, maybe 10 runs is not enough to get a stable result on ogbn-arxiv. (Note that once a initial seed is given, DGL can guarantee reproducibility, i.e. there is not randomness during training).

However, fixing the initial random leads to another kind of competition: people search for multiple random seeds and report the best test accuracy (though averaged for ten runs).

cc @Espylapiza , WDYT?

**Comment by yzh119** [comment]
btw, I recently noticed your paper "Combining Label Propagation and Simple Models Out-performs Graph Neural Networks", good job and it's quite impressive.

**Comment by Espylapiza** [comment]
Thanks for your question.
I'll check the implementation of the code.
The results came from several experiments with different parameters. The highest score is submitted, but I can't guarantee reproducibility.
I am considering to provide an option to guarantee reproducibility.

**Comment by Chillee** [comment]
@yzh119 

> As the initial seed is not fixed I think we can not guarantee reproducibility, maybe 10 runs is not enough to get a stable result on ogbn-arxiv.

I agree, however, at this point we have 30 runs (I just ran it again last night for 10 runs, and got `73.93 +- 0.14`, and running a 1-sample t-test I get a probability of `3e-8` that these samples come from a distribution with at least `73.99` mean.

So I suspect that the results submitted came from a different set of hyperparameters than what's listed.

Thanks for your comments on our paper :) I think the approach that DGL's submissions have been taking on ogbn-arxiv is also quite interesting - have you guys tried it on different datasets? I think there could potentially be some interesting findings if there's a general approach that could work well - unlike our approach, it might also work on ogbn-proteins :P

@Espylapiza 

> The results came from several experiments with different parameters. The highest score is submitted, but I can't guarantee reproducibility.

What do you mean by this?

**Comment by Chillee** [comment]
@yzh119 

> As the initial seed is not fixed I think we can not guarantee reproducibility, maybe 10 runs is not enough to get a stable result on ogbn-arxiv.

I agree, however, at this point we have 30 runs (I just ran it again last night for 10 runs, and got `73.93 +- 0.14`). So over our 30 runs the mean is `73.91` with `0.13` std.

and running a 1-sample t-test I get a probability of `0.001` that these samples come from a distribution with at least `73.99` mean.

So I suspect that the results submitted came from a different set of hyperparameters than what's listed.

Thanks for your comments on our paper :) I think the approach that DGL's submissions have been taking on ogbn-arxiv is also quite interesting - have you guys tried it on different datasets? I think there could potentially be some interesting findings if there's a general approach that could work well - unlike our approach, it might also work on ogbn-proteins :P

@Espylapiza 

> The results came from several experiments with different parameters. The highest score is submitted, but I can't guarantee reproducibility.

What do you mean by this?

**Comment by Espylapiza** [comment]
I mean the way I tune the hyperparameters might overestimate the score, because even using the same hyperparameters will have different results, and I submitted the highest one.

**Comment by Chillee** [comment]
I don't know the details of how you're doing hyperparameter search, but if there's a lot of noise in the training procedure then simply taking the best output from your hyperparameter search can have the effect of "optimizing your random seed". 

For example, say your only hyperparameter is some continuous value between `[0-1]` - it has a moderate effect on your final loss, but there's also just a lot of noise in your final test accuracy. If the optimal value is `0.5`, then your hyperparameter tuning might try `0.49`, `0.51`, `0.48`, ... Since such small changes in your hyperparameters won't affect the "true" test accuracy much, simply reporting the best accuracy from these choices is equivalent to rerunning your 10 run training procedure multiple times and selecting the best one.

**Comment by Espylapiza** [comment]
That's exactly what I mean.
I am trying to find a better set of hyperparameters to reach the score of 73.99 with a fixed random seed.
btw, the probablity is about 2.6% if your assumption are true:
```python3
import numpy as np

np.random.seed(0)
gen = lambda: np.random.normal(73.91, 0.13)
f = lambda n: np.mean([gen() for _ in range(n)])
cnt, total = 0, 100000
cnt = sum([1 if f(10) > 73.99 else 0 for _ in range(total)])
print(cnt / total)
```

**Comment by Chillee** [comment]
@Espylapiza I was using a sample of 30, from my 20 runs + @yzh119's reported 10 runs. Using your methodology, I get `0.00041`.

**Comment by Chillee** [comment]
Can you provide some more details on how the hyperparameters are searched?

**Comment by Chillee** [comment]
btw, irrelevant thing, but this line (https://github.com/Espylapiza/dgl/blob/master/examples/pytorch/ogb/ogbn-arxiv/gat.py#L58) should probably be `if args.use_labels`

**Comment by Espylapiza** [comment]
Thank you for pointint out the mistake!

The tuned hyperparameters are(in OGB's form):
`lr: [0.001, 0.002*, 0.003, 0.004, 0.005], n_layers: [3*], n_heads: [3*], n_hidden: [250*], dropout: [0.5, 0.6, 0.7, 0.75*, 0.8], input_drop: [0, 0.1, 0.2, 0.25*, 0.3, 0.4, 0.5], attn_drop: [0*], edge_drop: [0, 0.05, 0.1, 0.2, 0.3*, 0.4, 0.5], wd: [0*], mask_rate: [0.5*]`
These parameters are not searched on the final parameters, and with different random seed.

**Comment by Chillee** [comment]
@Espylapiza Any updates?

**Comment by Espylapiza** [comment]
I still need some time. Before that you can use 73.91 as a baseline.

**Comment by Espylapiza** [comment]
The result has been updated.

**Comment by Chillee** [comment]
@Espylapiza  thanks! BTW, would you accept a PR to add C&S on top of the existing GAT results? Adding C&S can get you to about 74.00 on arxiv, but submitting a result from us which only improves performance by ~0.1 seems somewhat excessive :P 

I think it would make more sense to just incorporate it into your existing submission :)

**Comment by Espylapiza** [comment]
Sure. Could you provide the code and results?

**Comment by Espylapiza** [comment]
No need, I can implement it by myself.

**Comment by Chillee** [comment]
One note is that when applying on top of these large GNN models (especially ones that are already using labels as features), you probably don't need the "correct" portion, and you only want to do the "smooth" portion.
